VMEncryption/volume-notif-svc/UdevVolNotif.c

```diff
@@ -0,0 +1,222 @@
+    if (pid > 0) {
+        custom_log("Parent process is exiting, pid=%d", pid);
+        exit(0);
+    }
+    umask(0);
+    pid_t sid = setsid();
+    if (sid < 0) {
+        custom_log("ERROR: Can't set sid (error no=%d)\n", errno);
+        exit(1);
+    }
+    if ((chdir("/")) < 0) {
+        custom_log("ERROR: Can't change directory (error no=%d)\n", errno);
+        exit(1);
+    }
+    close(STDIN_FILENO);
+    close(STDOUT_FILENO);
+    close(STDERR_FILENO);
+
+    //execve(argv[0], argv, NULL);
+    // char *newargv[] = { NULL };
+    // execv(argv[0], newargv);
+}
+
+int main(int argc, char *argv[]) {
+
+    prepare_log_file();
+
+    custom_log("\n\n### Azure Disk Encryption volume notification service ###");
+    custom_log("Starting udev volume notification program (%s) with %d args", argv[0], argc);
+    signal(SIGINT, sigint_handler);
+    custom_log("Registered SIGINT handler");
+
+    // Command line options: -d for daemon mode
+    int c = 0;
+    int daemon_mode = 0;
+    while ((c = getopt(argc, argv, "d")) != -1) {
+        switch (c) {
+            case 'd':
+                custom_log("Option for switching to daemon mode provided");
+                daemon_mode = 1;
+                break;
+            default:
+                custom_log("Unknown option %c", c);
+                exit(1);
+        }
+    }
+
+    if (daemon_mode) {
+        daemonize(argc, argv);
+    }
+
+    struct udev *udev;
+    struct udev_monitor *mon;
+    int fd;
+    int ret;
+
+    udev = udev_new();
+    if (!udev) {
+        custom_log("ERROR: Can't create udev (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    // Set custom log function
+
+    mon = udev_monitor_new_from_netlink(udev, "udev");
+    if (!mon) {
+        custom_log("ERROR: Can't create udev monitor (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    ret = udev_monitor_filter_add_match_subsystem_devtype(mon, "block", "partition" /*"disk"*/);
+    if (ret < 0) {
+        custom_log("ERROR: Can't add udev monitor filter (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    ret = udev_monitor_enable_receiving(mon);
+    if (ret < 0) {
+        custom_log("ERROR: Can't enable udev monitor receiving (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    fd = udev_monitor_get_fd(mon);
+    if (fd < 0) {
+        custom_log("ERROR: Can't get udev monitor fd (error no=%d)\n", errno);
+        exit(1);
+    }
+
+    custom_log("Entering main event loop");
+    while (1) {
+        fd_set fds;
+        struct timeval tv;
+        int ret;
+
+        FD_ZERO(&fds);
+        FD_SET(fd, &fds);
+        tv.tv_sec = 0;
+        tv.tv_usec = 0;
+
+        ret = select(fd+1, &fds, NULL, NULL, &tv);
+
+        if (ret > 0 && FD_ISSET(fd, &fds)) {
+            struct udev_device *dev;
+
+            dev = udev_monitor_receive_device(mon);
+            custom_log("A new udev monitoring event is received!");
+
+            const char *action = udev_device_get_action(dev);
```

Comment:
null pointer checks needed for all the char* pointers which are returned before using them